library/blob.c

```diff
@@ -2182,23 +2216,29 @@ int eblob_plain_writev(struct eblob_backend *b, struct eblob_key *key,
 			goto err_out_unlock;
 		prepared = 1;
 	}
+	pthread_mutex_unlock(&b->lock);
 
 	wc.flags = flags;
 	if (b->cfg.blob_flags & EBLOB_NO_FOOTER)
 		wc.flags |= BLOB_DISK_CTL_NOCSUM;
 	err = eblob_writev_raw(key, &wc, iov, iovcnt);
 	if (err)
-		goto err_out_unlock;
+		goto err_out_cleanup_wc;
 
 	/* Re-commit record to ram if it was copied */
 	if (prepared) {
 		err = eblob_commit_ram(b, key, &wc);
 		if (err != 0)
-			goto err_out_unlock;
+			goto err_out_cleanup_wc;
 	}
 
+	goto err_out_cleanup_wc;
+
 err_out_unlock:
 	pthread_mutex_unlock(&b->lock);
```

Comment:
Exactly this is called goto-hell.

You should make error path always going through correct set of labels, it is not allowed to have unlock somewhere in the middle of the function and set of gotos to unlock. Also, making 'goto err_out_cleanup_wc' in the middle of the correct (i.e. not error) path is a red sign.

Please clean this up.
There should be exactly one error path, all error handlers should goto to appropriate labels with the correct order.
Non-error path can pass through error labels if error is zero.